### PR TITLE
Add Backend class [ADAM-107] 

### PR DESCRIPTION
**Pull request by moeyensj**
This PR is branched off of #7, so requires that to be merged first. 

This PR introduces a new "Backend" class that has several abstract methods that need to be defined by subclasses. Once implemented, the Backend class can then distribute those methods as multiple processes per chunk of orbits. 

One important change is that we make PYOORB the default integrator that comes with `adam_core` which I think is reasonable.  

Here is an example from the README:
```python
import numpy as np
from astropy.time import Time
from astropy import units as u
from adam_core.orbits.query import query_horizons
from adam_core.backend import PYOORB

# Get orbit to propagate
initial_time = Time([60000.0], scale="tdb", format="mjd")
object_ids = ["Duende", "Eros", "Ceres"]
orbits = query_horizons(object_ids, initial_time)

# Make sure PYOORB is ready
backend = PYOORB()

# Define propagation times
times = initial_time + np.arange(0, 100) * u.d

# Propagate orbits! This function supports multiprocessing for large
# propagation jobs. 
propagated_orbits = backend.propagate_orbits(
    orbits,
    times,
    chunk_size=100,
    num_jobs=1,
)
```

Additional action items to be addressed in future PRs:
- [ ] Add observers class 
  - [ ] Add spiceypy utility to get observer coordinates
  - [ ] Add code to manage SPICE "kernel" files
- [ ] Add ephemeris class
- [ ] Add PYOORB ephemeris generation

### CODE REVIEW

File: adam_core/backend/backend.py
Comment (spenczar) [inline adam_core/backend/backend.py:78]: Why does the backend need to be copied?

Also - probably worth loudly documenting that backends need to be picklable. `multiprocessing` uses `pickle` to send these things to the other processes. That means that backends can't, for example, have any open file handles or network connections.

File: adam_core/backend/pyoorb.py
Comment (spenczar) [inline adam_core/backend/pyoorb.py:49]: Have you confirmed that this works with multiprocessing? `__init__` usually isn't run when unpickling a class instance, so I'd think that this `oorb_init` call would only happen in the main process and not the pool's subprocesses. I could be totally wrong though.

File: adam_core/backend/backend.py
Comment (spenczar) [inline adam_core/backend/backend.py:107]: Please annotate the type of `observers`

File: adam_core/backend/pyoorb.py
Comment (spenczar) [inline adam_core/backend/pyoorb.py:33]: Why not the more standard way of setting optional kwargs with defaults?

```suggestion
    def __init__(self, *, dynamical_model: str="N", ephemeris_file: str = "de430.dat"):
```

File: adam_core/backend/pyoorb.py
Comment (spenczar) [inline adam_core/backend/pyoorb.py:40]: What's the purpose behind this part?

File: adam_core/backend/pyoorb.py
Comment (spenczar) [inline adam_core/backend/pyoorb.py:44]: Please check if CONDA_PREFIX is in env vars first

File: adam_core/backend/pyoorb.py
Comment (spenczar) [reply adam_core/backend/pyoorb.py:40]: Oh, I just read further and figured it out.

I still don't really understand, though! So this is to make sure you call oorb_init just once per process tree? Why?

I propose naming the env var something like `ADAM_CORE_PYOORB_INITIALIZED` to be a bit clearer.

File: adam_core/backend/pyoorb.py
Comment (spenczar) [inline adam_core/backend/pyoorb.py:210]:
```suggestion
        epochs_pyoorb : `~numpy.ndarray` (N, 2)
```

File: adam_core/backend/pyoorb.py
Comment (spenczar) [inline adam_core/backend/pyoorb.py:147]: I think it's nice to use [`enum`](https://docs.python.org/3/library/enum.html) for things like this. You'd get type checking and the code will end up being a lot less verbose. 

```python
class OpenOrbTimescale(enum.Enum):
    UTC = 1
    UT1 = 2
    TT = 3
    TAI = 4

...

def _configure_orbits(
        orbits: np.ndarray,
        t0: np.ndarray,
        orbit_type: str,
        time_scale: OpenOrbTimescale,
        magnitude: Optional[float] = None,
        slope: Optional[float] = None,
    ) -> np.ndarray:

        ...
        time_scale_list = [time_scale.value for i in range(num_orbits)]   
```

Called like `PYOORB._configure_orbits(..., time_scale=OpenOrbTimescale.UTC, ...)`

File: adam_core/backend/pyoorb.py
Comment (spenczar) [inline adam_core/backend/pyoorb.py:157]: Looks like slope and magnitude are ignored if they are `ndarray`s, which doesn't seem right.

File: adam_core/backend/pyoorb.py
Comment (spenczar) [inline adam_core/backend/pyoorb.py:97]: What's the advantage of being able to pass in the keplerian and cometary representations to openorb, instead of always using cartesian?

File: adam_core/backend/pyoorb.py
Comment (spenczar) [inline adam_core/backend/pyoorb.py:292]: This seems a bit surprising - why always convert to tdb?

File: adam_core/utils/multiprocessing.py
Comment (spenczar) [inline adam_core/utils/multiprocessing.py:20]: `enable_parallel` doesn't provide a lot more over returning just `num_workers`. Callers could check if `num_workers > 1`. I'd recommend trimming that down, I think it will feel simpler.

Also I'm not convinced that this adds much value in general, yeah. Callers could just use `mp.cpu_count()` themselves, this API is a bit indirect.

File: adam_core/backend/backend.py
Comment (spenczar) [inline adam_core/backend/backend.py:26]: I think `Backend` is too generic of a name, and risks getting overloaded with tons of stuff. I'd recommend naming this `Propagator`. Ephemeris generation stands out as a little bit extra here (add an `EphemerisGenerator` ABC perhaps?) but I think that's acceptable enough.

**Review by spenczar — CHANGES_REQUESTED** [review]
Agreed that `_check_parallel` can be discarded. Let callers set the number of workers to 1 or to `cpu_count()` themselves. You can disable using the multiprocessing pool when workers=1, I don't think the `_check_parallel` API is necessary for that.

The basic structure looks alright, but I think it could benefit from trimming down. Do we need to accept Keplerian orbital parameters in internal function calls, for example? I think shedding a lot of that weight would make this easier to read and maintain.

File: adam_core/backend/pyoorb.py
Comment (moeyensj) [reply adam_core/backend/pyoorb.py:292]: This is to match what JPL does: by default: orbits in TDB, observations in UTC

File: adam_core/backend/backend.py
Comment (moeyensj) [reply adam_core/backend/backend.py:78]: I removed the copies and added calls to `itertools.repeat` rather than list comprehensions. Hopefully that is cleaner.

Also added a statement that subclasses must be pickleable.

File: adam_core/backend/pyoorb.py
Comment (moeyensj) [reply adam_core/backend/pyoorb.py:28]: If you call `oorb_init` twice from the same process it will return an error code 1. So this is an attempt to avoid that.

Comment (moeyensj) [reply adam_core/backend/pyoorb.py:28]: Also, renamed the env var as you suggested.

File: adam_core/backend/pyoorb.py
Comment (moeyensj) [reply adam_core/backend/pyoorb.py:32]: Done!

File: adam_core/backend/pyoorb.py
Comment (moeyensj) [reply adam_core/backend/pyoorb.py:198]: Fixed.

File: adam_core/backend/pyoorb.py
Comment (moeyensj) [reply adam_core/backend/pyoorb.py:135]: I've added two enums (one for orbit types and the other for time scales).

File: adam_core/backend/pyoorb.py
Comment (moeyensj) [reply adam_core/backend/pyoorb.py:85]: Ah so this comes from the olden days before I decided to go full cartesian. That said, we know openorb does not treat all representations equally so I've kept this in the event we ever want to try different representations. Happy to remove it if you want me to.

File: adam_core/backend/backend.py
Comment (moeyensj) [reply adam_core/backend/backend.py:26]: Renamed to `Propagator`. I'd like to keep ephemeris generation as part of the same class that does propagation because you do both of those together so frequently.

File: adam_core/backend/pyoorb.py
Comment (moeyensj) [reply adam_core/backend/pyoorb.py:37]: Tested and confirmed that multiprocessing works!

File: adam_core/backend/pyoorb.py
Comment (moeyensj) [reply adam_core/backend/pyoorb.py:145]: Fixed. I actually re-did the openorb orbit formatting completely and now create a correctly formatted empty 2D  array rather which is then filled with the correct values. I think its miles better than this uglier format which was drawn from the pyoorb README.

File: adam_core/utils/multiprocessing.py
Comment (moeyensj) [reply adam_core/utils/multiprocessing.py:20]: Removed, turns out the `mp.Pool` accepts both an int and None. If its None then it will use `mp.cpu_count()` processes.

File: adam_core/backend/backend.py
Comment (moeyensj) [reply adam_core/backend/backend.py]: I'm going to hold off on this until I've defined the Observer's class in a follow up PR.

**Review by spenczar — APPROVED** [review]
Way improved. Nice!

**Comment by moeyensj** [comment]
This PR will be rebased and modified once #17 is merged.

**Pull request merged**
